### PR TITLE
[ADD] protect timesheet lines with leaves from deletion

### PR DESCRIPTION
**Pull request by hbrunn**

### CODE REVIEW

File: hr_timesheet_holiday/models/hr_analytic_timesheet.py [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
     leave_id = fields.Many2one('hr.holidays', 'Leave id')
+
+    @api.multi
+    def unlink(self):
+        leaves = self.mapped('leave_id')
+        if leaves and set(leaves.mapped('state')) - set([
+                'draft', 'cancel', 'refuse',
+        ]):
+            raise UserError(_(
+                'You can\'t delete timesheet lines with active leaves',
```
Comment (nikul-serpentcs) [inline hr_timesheet_holiday/models/hr_analytic_timesheet.py]: `\'t` ?? No need

Comment (hbrunn) [reply hr_timesheet_holiday/models/hr_analytic_timesheet.py]: that's an escaped quote

Comment (nikul-serpentcs) [reply hr_timesheet_holiday/models/hr_analytic_timesheet.py]: @hbrunn Change message like this "You can't delete timesheet lines with active leaves "
SO no need to escaped quote

**Review by thomaspaulb — APPROVED** [review]
Very fancy use of sets, and supported by tests.

Travis is not green, but I think that is not because of the code.

File: hr_timesheet_holiday/models/hr_analytic_timesheet.py (hunk 1 above)
Comment (thomaspaulb) [reply hr_timesheet_holiday/models/hr_analytic_timesheet.py]: I do prefer that also, but I don't think this is of breaking importance.

**Comment by alexey-pelykh** [comment]
Related to https://github.com/OCA/timesheet/pull/175

File: hr_timesheet_holiday/tests/test_all.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 import time
 from openerp.tests.common import TransactionCase
-from openerp.exceptions import ValidationError
+from openerp.exceptions import UserError, ValidationError
```
Comment (thomaspaulb) [inline hr_timesheet_holiday/tests/test_all.py]: @hbrunn This should be `Warning as UserError`

File: hr_timesheet_holiday/tests/test_all.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 import time
 from openerp.tests.common import TransactionCase
-from openerp.exceptions import UserError, ValidationError
+from openerp.exceptions import Warning as UserError, ValidationError
```
Comment (alexey-pelykh) [inline hr_timesheet_holiday/tests/test_all.py:6]: Seems like a voodoo magic, why not to change all UserError to Warning?

Comment (thomaspaulb) [reply hr_timesheet_holiday/tests/test_all.py:6]: [OCA guidelines](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#33idioms) point 7: Exceptions.

Comment (hbrunn) [reply hr_timesheet_holiday/tests/test_all.py:6]: @alexey-pelykh that's a wrong choice of class name in earlier versions. You missed all the fun back then ;-)

Comment (hbrunn) [reply hr_timesheet_holiday/tests/test_all.py:6]: (the reason it was a wrong choice is that it masks https://docs.python.org/2/library/exceptions.html#exceptions.Warning)

Comment (alexey-pelykh) [reply hr_timesheet_holiday/tests/test_all.py:6]: Yep, missed all the mysterious fun :) 👍

**Review by alexey-pelykh — APPROVED** [review]
Code seems legit

**Comment by pedrobaeza** [comment]
@hbrunn please remember to use the ocabot for merging the PRs, that also allows to auto-bump the module versions, and performs other changes in the same batch (translations, README, etc).

**Comment by hbrunn** [comment]
whoops, you're right. Will do it the right way too for old versions, it's just that most of the 8.0 branches fail, so I don't trust ocabot there

**Comment by pedrobaeza** [comment]
Yeah, that's true, and ocabot won't merge it if fails, but the final objective would be to fix the branch. I agree that's sometimes not possible, so I let it to your choice. Just remember that the option is there and has other advantages.

**Pull request merged**